azurerm/resource_arm_application_insights_api_key.go

```diff
@@ -0,0 +1,186 @@
+	client := meta.(*ArmClient).appInsightsAPIKeyClient
+	ctx := meta.(*ArmClient).StopContext
+
+	log.Printf("[INFO] preparing arguments for AzureRM Application Insights API key creation.")
+
+	name := d.Get("name").(string)
+	appInsightsID := d.Get("application_insights_id").(string)
+
+	id, err := parseAzureResourceID(appInsightsID)
+	if err != nil {
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	appInsightsName := id.Path["components"]
+
+	if requireResourcesToBeImported && d.IsNewResource() {
+		var existing insights.ApplicationInsightsComponentAPIKey
+		existing, err = client.Get(ctx, resGroup, appInsightsName, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for presence of existing Application Insights API key %q (Resource Group %q): %s", name, resGroup, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_application_insights_api_key", *existing.ID)
+		}
+	}
+
+	apiKeyProperties := insights.APIKeyRequest{
+		Name:                  &name,
+		LinkedReadProperties:  azure.ExpandApplicationInsightsAPIKeyLinkedProperties(d.Get("read_permissions").(*schema.Set), appInsightsID),
+		LinkedWriteProperties: azure.ExpandApplicationInsightsAPIKeyLinkedProperties(d.Get("write_permissions").(*schema.Set), appInsightsID),
+	}
+
+	result, err := client.Create(ctx, resGroup, appInsightsName, apiKeyProperties)
+	if err != nil {
+		return fmt.Errorf("Error creating Application Insights API key %q (Resource Group %q): %+v", name, resGroup, err)
+	}
+
+	if result.APIKey == nil {
+		return fmt.Errorf("Error creating Application Insights API key %q (Resource Group %q): got empty API key", name, resGroup)
+	}
+
+	d.SetId(*result.ID)
+
+	// API key can only retrieved at key creation
+	d.Set("api_key", result.APIKey)
+
+	return resourceArmApplicationInsightsAPIKeyRead(d, meta)
+}
+
+func resourceArmApplicationInsightsAPIKeyRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).appInsightsAPIKeyClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Reading AzureRM Application Insights API key '%s'", id)
+
+	resGroup := id.ResourceGroup
+	appInsightsName := id.Path["components"]
+	keyID := id.Path["apikeys"]
+
+	result, err := client.Get(ctx, resGroup, appInsightsName, keyID)
+	if err != nil {
+		if utils.ResponseWasNotFound(result.Response) {
+			log.Printf("[WARN] AzureRM Application Insights API key '%s' not found, removing from state", id)
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on AzureRM Application Insights API key '%s': %+v", keyID, err)
+	}
+
+	d.Set("application_insights_id", fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/microsoft.insights/components/%s", client.SubscriptionID, resGroup, appInsightsName))
```

Comment:
Why are we regenerating and setting this when it's a required part of the schema? If this is any different then the one the user supplied wouldn't we have a diff and a new resource be created?

Reply:
Good point!

Reply:
On second thought, cannot the diff be caught only if the state is set with the actual value returned by the API?

Also, I believe it is not uncommon to import resources with mostly empty resource definitions then use the output of the next plan to complete the configurations.

What do you think?

Reply:
Removing this line breaks the acceptance tests:

```
# make testacc TEST=./azurerm TESTARGS='-run=TestAccAzureRMApplicationInsightsAPIKey_.* -count=1'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -run=TestAccAzureRMApplicationInsightsAPIKey_.* -count=1 -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMApplicationInsightsAPIKey_no_permission
=== PAUSE TestAccAzureRMApplicationInsightsAPIKey_no_permission
=== RUN   TestAccAzureRMApplicationInsightsAPIKey_requiresImport
--- SKIP: TestAccAzureRMApplicationInsightsAPIKey_requiresImport (0.00s)
    resource_arm_application_insights_api_key_test.go:33: Skipping since resources aren't required to be imported
=== RUN   TestAccAzureRMApplicationInsightsAPIKey_read_telemetry_permissions
=== PAUSE TestAccAzureRMApplicationInsightsAPIKey_read_telemetry_permissions
=== RUN   TestAccAzureRMApplicationInsightsAPIKey_write_annotations_permission
=== PAUSE TestAccAzureRMApplicationInsightsAPIKey_write_annotations_permission
=== RUN   TestAccAzureRMApplicationInsightsAPIKey_authenticate_permission
=== PAUSE TestAccAzureRMApplicationInsightsAPIKey_authenticate_permission
=== RUN   TestAccAzureRMApplicationInsightsAPIKey_full_permissions
=== PAUSE TestAccAzureRMApplicationInsightsAPIKey_full_permissions
=== CONT  TestAccAzureRMApplicationInsightsAPIKey_no_permission
=== CONT  TestAccAzureRMApplicationInsightsAPIKey_authenticate_permission
=== CONT  TestAccAzureRMApplicationInsightsAPIKey_write_annotations_permission
=== CONT  TestAccAzureRMApplicationInsightsAPIKey_full_permissions
=== CONT  TestAccAzureRMApplicationInsightsAPIKey_read_telemetry_permissions
--- PASS: TestAccAzureRMApplicationInsightsAPIKey_no_permission (65.68s)
--- FAIL: TestAccAzureRMApplicationInsightsAPIKey_write_annotations_permission (79.87s)
    testing.go:538: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.

        (map[string]string) {
        }


        (map[string]string) (len=1) {
         (string) (len=23) "application_insights_id": (string) (len=175) "/subscriptions/499bbc6c-d927-488d-8e6d-a463319d78fe/resourceGroups/acctestRG-1733103138601821768/providers/microsoft.insights/components/acctestappinsights-1733103138601821768"
        }

--- FAIL: TestAccAzureRMApplicationInsightsAPIKey_read_telemetry_permissions (80.01s)
    testing.go:538: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.

        (map[string]string) {
        }


        (map[string]string) (len=1) {
         (string) (len=23) "application_insights_id": (string) (len=175) "/subscriptions/499bbc6c-d927-488d-8e6d-a463319d78fe/resourceGroups/acctestRG-6632056159759859543/providers/microsoft.insights/components/acctestappinsights-6632056159759859543"
        }

--- FAIL: TestAccAzureRMApplicationInsightsAPIKey_full_permissions (82.20s)
    testing.go:538: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.

        (map[string]string) {
        }


        (map[string]string) (len=1) {
         (string) (len=23) "application_insights_id": (string) (len=175) "/subscriptions/499bbc6c-d927-488d-8e6d-a463319d78fe/resourceGroups/acctestRG-5282038476415038603/providers/microsoft.insights/components/acctestappinsights-5282038476415038603"
        }

--- FAIL: TestAccAzureRMApplicationInsightsAPIKey_authenticate_permission (82.34s)
    testing.go:538: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.

        (map[string]string) {
        }


        (map[string]string) (len=1) {
         (string) (len=23) "application_insights_id": (string) (len=175) "/subscriptions/499bbc6c-d927-488d-8e6d-a463319d78fe/resourceGroups/acctestRG-6680669483478120498/providers/microsoft.insights/components/acctestappinsights-6680669483478120498"
        }

FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm       82.365s
make: *** [GNUmakefile:23: testacc] Error 1
```